File: CopyPaste.Core/CleanupService.cs
namespace CopyPaste.Core;

/// <summary>
/// Service responsible for automatic cleanup of old clipboard items.
/// Runs once per day, checking every 18 hours.
/// </summary>
public sealed class CleanupService : IDisposable
{
    private static readonly TimeSpan _checkInterval = TimeSpan.FromHours(18);
    private const string _lastCleanupFileName = "last_cleanup.txt";

    private readonly IClipboardRepository _repository;
    private readonly Func<int> _getRetentionDays;
    private readonly Timer _timer;
    private bool _isDisposed;

    private static string CleanupFilePath => Path.Combine(
        Path.GetDirectoryName(StorageConfig.DatabasePath) ?? string.Empty,
        _lastCleanupFileName
    );

    public CleanupService(IClipboardRepository repository, Func<int> getRetentionDays, bool startTimer = true)
    {
        ArgumentNullException.ThrowIfNull(repository);
        ArgumentNullException.ThrowIfNull(getRetentionDays);

        _repository = repository;
        _getRetentionDays = getRetentionDays;

        _timer = new Timer(
            _ => RunCleanupIfNeeded(),
            null,
            startTimer ? TimeSpan.Zero : Timeout.InfiniteTimeSpan,
            startTimer ? _checkInterval : Timeout.InfiniteTimeSpan
        );
    }

    internal void RunCleanupIfNeeded()
    {
        if (_isDisposed) return;

        int retentionDays = _getRetentionDays();
        if (retentionDays <= 0) return;

        var lastCleanupDate = LoadLastCleanupDate();
        if (lastCleanupDate.Date == DateTime.UtcNow.Date) return;

        try
        {
            int deletedCount = _repository.ClearOldItems(retentionDays, excludePinned: true);
            SaveLastCleanupDate(DateTime.UtcNow);

            if (deletedCount > 0)
            {
                AppLogger.Info($"Cleanup: {deletedCount} items older than {retentionDays} days removed");
            }
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            AppLogger.Exception(ex, "Cleanup failed");
        }
    }

    private static DateTime LoadLastCleanupDate()
    {
        try
        {
            if (File.Exists(CleanupFilePath))
            {
                string content = File.ReadAllText(CleanupFilePath);
                if (DateTime.TryParse(content, out var date))
                {
                    return date;
                }
            }
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            AppLogger.Exception(ex, "Failed to load last cleanup date");
        }

        return DateTime.MinValue;
    }

    private static void SaveLastCleanupDate(DateTime date)
    {
        try
        {
            var directory = Path.GetDirectoryName(CleanupFilePath);
            if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
            {
                Directory.CreateDirectory(directory);
            }
            File.WriteAllText(CleanupFilePath, date.ToString("O"));
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            AppLogger.Exception(ex, "Failed to save last cleanup date");
        }
    }

    public void Dispose()
    {
        if (_isDisposed) return;
        _timer.Dispose();
        _isDisposed = true;
    }
}

File: v2/core/lib/services/cleanup_service.dart
import 'dart:async';
import 'dart:io';

import 'package:path/path.dart' as p;

import '../config/storage_config.dart';
import '../repository/i_clipboard_repository.dart';

class CleanupService {
  CleanupService(this._repository, this._getRetentionDays, {StorageConfig? storage})
      : _storage = storage;

  static const Duration _checkInterval = Duration(hours: 18);
  static const String _lastCleanupFileName = 'last_cleanup.txt';

  final IClipboardRepository _repository;
  final int Function() _getRetentionDays;
  final StorageConfig? _storage;
  Timer? _timer;
  bool _disposed = false;

  String? _baseDirPath;

  String get _cleanupFilePath =>
      p.join(_baseDirPath ?? '', _lastCleanupFileName);

  void start(String baseDirPath) {
    _baseDirPath = baseDirPath;
    _timer = Timer.periodic(_checkInterval, (_) => runCleanupIfNeeded());
    runCleanupIfNeeded();
  }

  Future<void> runCleanupIfNeeded() async {
    if (_disposed) return;
    final retentionDays = _getRetentionDays();
    if (retentionDays <= 0) return;

    final lastCleanup = _loadLastCleanupDate();
    final now = DateTime.now().toUtc();
    if (lastCleanup.year == now.year &&
        lastCleanup.month == now.month &&
        lastCleanup.day == now.day) {
      return;
    }

    try {
      await _repository.clearOldItems(retentionDays, excludePinned: true);
      _saveLastCleanupDate(now);
      await _cleanOrphanImages();
    } catch (_) {}
  }

  DateTime _loadLastCleanupDate() {
    try {
      final file = File(_cleanupFilePath);
      if (file.existsSync()) {
        final content = file.readAsStringSync().trim();
        final parsed = DateTime.tryParse(content);
        if (parsed != null) return parsed.toUtc();
      }
    } catch (_) {}
    return DateTime.utc(2000);
  }

  void _saveLastCleanupDate(DateTime date) {
    try {
      final file = File(_cleanupFilePath);
      file.parent.createSync(recursive: true);
      file.writeAsStringSync(date.toIso8601String());
    } catch (_) {}
  }

  void dispose() {
    _disposed = true;
    _timer?.cancel();
  }

  Future<void> _cleanOrphanImages() async {
    final storage = _storage;
    if (storage == null) return;
    try {
      final validPaths = await _repository.getImagePaths();
      storage.cleanOrphanImages(validPaths);
    } catch (_) {}
  }
}
